File: src/main/java/com/scottscmo/ui/panels/SettingsPanel.java
package com.scottscmo.ui.panels;

import com.scottscmo.AppLogger;
import com.scottscmo.Config;
import com.scottscmo.ui.components.FileEditor;
import com.scottscmo.ui.components.Form;
import com.scottscmo.ui.components.FormInput;
import com.scottscmo.util.Cryptor;
import net.miginfocom.swing.MigLayout;

import javax.swing.*;
import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.security.GeneralSecurityException;
import java.util.Map;

public final class SettingsPanel extends JPanel {
    public SettingsPanel() {
        // config editor
        var configEditor = new FileEditor(Path.of(Config.CONFIG_PATH), "", 15, 55);
        var configReloadButton = new JButton("Reload Configurations");

        // api importer
        var defaultCredentialsPath = Config.getRelativePath(Config.GOOGLE_API_DIR + "/credentials.json");
        var googleApiCredentialImportForm = new Form("Google API Credentials Importer",
                Map.of(
                        "credentialsFilePath", new FormInput("Credentials json", "file", defaultCredentialsPath)
                ),
                form -> {
                    storeGoogleAPICredentials(form.getValue("credentialsFilePath"));
                    return "Credentials imported! Please restart app.";
                }
        );

        setLayout(new MigLayout("", "[]10[]", "top"));
        var configEditorContainer = new JPanel();
        configEditorContainer.setLayout(new MigLayout("ins 0"));
        configEditorContainer.add(configEditor.getUI(), "wrap");
        configEditorContainer.add(configReloadButton);
        add(new BibleInfoPanel(), "wrap");
        add(googleApiCredentialImportForm.getUI(), "wrap");
        add(configEditorContainer);

        // controls
        configReloadButton.addActionListener(e -> {
            Config.reload();
        });
    }

    private void storeGoogleAPICredentials(String credentialsFilePath) {
        // encrypt credentials
        try {
            Cryptor.encryptFile(credentialsFilePath,
                    Config.getRelativePath(Config.GOOGLE_API_CREDENTIALS_PATH),
                    Config.get().clientInfoKey()
            );
        } catch (GeneralSecurityException e) {
            AppLogger.showError("Failed to encrypt Google API key", e);
            return;
        } catch (IOException e) {
            AppLogger.showError("Failed to load Google API credentials", e);
            return;
        }

        // remove existing token
        var storedToken = new File(Config.getRelativePath(Config.GOOGLE_API_DIR + "/StoredCredential"));
        if (storedToken.exists()) {
            storedToken.delete();
        }
    }
}

File: src/main/kotlin/com/scottscmo/ui/panels/SettingsPanel.kt
package com.scottscmo.ui.panels

import com.scottscmo.Config
import com.scottscmo.ui.components.FileEditor
import com.scottscmo.ui.components.Form
import com.scottscmo.ui.components.FormInput
import com.scottscmo.util.Cryptor
import net.miginfocom.swing.MigLayout
import java.nio.file.Path
import javax.swing.JButton
import javax.swing.JPanel

class SettingsPanel : JPanel() {
    init {
        layout = MigLayout("", "[]10[]", "top")

        // config editor
        val configEditor = FileEditor(Path.of(Config.CONFIG_PATH), "", 15, 55)
        add(JPanel().apply {
            layout = MigLayout("ins 0")
            add(configEditor.ui, "wrap")
            add(JButton("Re-Apply Configurations").apply {
                addActionListener {
                    Config.load()
                }
            })
        })

        // api importer
        add(Form("Google API Credentials Importer", mapOf(
            "credentialsFilePath" to FormInput("Credentials json", "file",
                Config.getRelativePath("${Config.GOOGLE_API_DIR}/credentials.json")),
        )) {
            require(Config.get().clientInfoKey.isNotEmpty()) { "clientInfoKey is missing from config.yaml!" }

            Cryptor.encryptFile(it["credentialsFilePath"],
                Config.getRelativePath(Config.GOOGLE_API_CREDENTIALS_PATH),
                Config.get().clientInfoKey)
            "Credentials imported!"
        }.ui)
    }
}